// @meshconnect/web-link-sdk/utils/wallet/EVMWalletStrategy.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { BaseWalletStrategy } from './WalletStrategy';
import { connectToEVMWallet, disconnectFromEVMWallet, signEVMMessage, sendEVMTransaction, sendEVMTokenTransaction, switchEVMChain, findAvailableProviders } from '../connectors/evm';
var EVMWalletStrategy = /** @class */ (function (_super) {
    __extends(EVMWalletStrategy, _super);
    function EVMWalletStrategy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EVMWalletStrategy.prototype.connect = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, connectToEVMWallet(payload.integrationName, payload.targetChainId ? parseInt(payload.targetChainId, 10) : undefined)];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [2 /*return*/, {
                                accounts: result.accounts,
                                chainId: result.chainId,
                                isConnected: result.isConnected
                            }];
                    case 2:
                        error_1 = _a.sent();
                        throw this.handleError(error_1, 'connect to EVM wallet');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.disconnect = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, disconnectFromEVMWallet(payload.walletName || 'Unknown Wallet')];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        error_2 = _a.sent();
                        throw this.handleError(error_2, 'disconnect from EVM wallet');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.signMessage = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, signEVMMessage(payload.walletName || 'Unknown Wallet', payload.address, payload.message)];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [2 /*return*/, result];
                    case 2:
                        error_3 = _a.sent();
                        throw this.handleError(error_3, 'sign EVM message');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.switchChain = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, switchEVMChain(payload.chainId)];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [2 /*return*/, result];
                    case 2:
                        error_4 = _a.sent();
                        throw this.handleError(error_4, 'switch EVM chain');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.sendNativeTransfer = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, sendEVMTransaction(payload.toAddress, BigInt(payload.amount * Math.pow(10, payload.decimalPlaces)), payload.account, payload.gasLimit, payload.maxFeePerGas, payload.maxPriorityFeePerGas)];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [2 /*return*/, result];
                    case 2:
                        error_5 = _a.sent();
                        throw this.handleError(error_5, 'send EVM native transfer');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.sendSmartContractInteraction = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var result, error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, sendEVMTokenTransaction(payload.address, JSON.parse(payload.abi), payload.functionName, payload.args, payload.account)];
                    case 1:
                        result = _a.sent();
                        if (result instanceof Error) {
                            throw result;
                        }
                        return [2 /*return*/, result];
                    case 2:
                        error_6 = _a.sent();
                        throw this.handleError(error_6, 'send EVM smart contract interaction');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EVMWalletStrategy.prototype.getProviders = function () {
        return findAvailableProviders().map(function (provider) { return ({
            icon: provider.icon,
            id: provider.id,
            name: provider.name,
            type: 'evm'
        }); });
    };
    return EVMWalletStrategy;
}(BaseWalletStrategy));
export { EVMWalletStrategy };
